### ISSUE TITLE
Intermittent 600 s hang of test_issue_29 on macos-15/py3.13 kills the xdist worker and ~20 following tests

### ISSUE THREAD

**Issue by Josef-Haupt** [birdnet-team/birdnet#68]
## Symptom

`src/birdnet_tests/issues/test_issue29.py::test_issue_29` intermittently hangs on the `macos-15` / Python 3.13 lane until the 600 s pytest-timeout backstop kills the xdist worker. Because the thread timeout method kills the whole process, this surfaces as the opaque `[gw0] node down: Not properly terminated`, and with `--max-worker-restart=0` the ~20 tests remaining on that worker are abandoned, so the whole job goes red.

Seen twice so far, identically (both on gw0, both at ~97%):

- `main` at `da4cf87` (post-#66): https://github.com/birdnet-team/birdnet/actions/runs/31533781263
- PR #67 branch at `ce047f7`: https://github.com/birdnet-team/birdnet/actions/runs/31536419236

It passed on the same lane in the runs before and after (e.g. https://github.com/birdnet-team/birdnet/actions/runs/31582249722), so it is a race, present both with and without #67.

## It is a wedge, not slowness

In the passing post-#67 run the test completed in about **1 second** on a warm worker (logstart 09:24:54.83 → PASSED 09:24:55.82). The failures sit at exactly the 600 s backstop. A 600× gap is a deadlock/stall, not the tail of a slow distribution, so raising the timeout would not help.

## Why there is no stack trace, and why CI cannot currently produce one

Both failing runs uploaded watchdog dumps, and in both the gw0 file ends with

```
===== STARTING: src/birdnet_tests/issues/test_issue29.py::test_issue_29 =====
```

and nothing after. The instrumentation loses this hang structurally: pytest-timeout (600 s) always fires before the faulthandler watchdog (test timeout + 60 s), and the thread-method's own stack dump goes to the xdist worker's stderr, which dies with the worker. So the two existing evidence channels cancel each other out for in-test hangs under xdist.

## What is already ruled out (from the #67 discussion)

- **Package drift** — a passing and both failing jobs installed byte-identical package sets (TF 2.21.0, keras 3.15.1, numpy 2.5.2, h5py 3.14.0, soundfile 0.14.0, litert 2.1.6).
- **#66** — the test got *faster* across the merge boundary locally (6.5–7.3 s → 5.5 s), as expected from removing the poll barrier.
- **#67** — the second failure includes #67's fix, the first predates it.

## Why this test is a plausible victim

`test_issue_29` replays the reporter's exact config from #29 and is the **only test in the suite** that passes `n_workers=None`, which resolves to one spawned worker process per physical core — each importing TensorFlow — plus `show_stats="progress"`, which additionally starts the performance-tracker process. On a 3-core / 7 GB macOS runner, at ~97% through the run, with two sibling xdist workers concurrently running heavy model tests, that is a spawn-import stampede under peak memory pressure. The pipeline parent waits on its children with `timeout=None` (`finish_signal.wait` / `q.get`), so one child stalling during startup wedges the test silently — which matches the total absence of output after logstart.

That is a hypothesis, not a proven cause; nothing above pins the stuck frame.

## Options

1. **Pin `n_workers=1` in the test.** The regression #29 actually guards (the progress-dispatcher teardown assert, `show_stats="progress"`) is independent of worker count, and the `n_workers=None` resolution logic is separately unit-tested in `configs_py/ProcessingConfig/test_validate_n_workers.py` — so net coverage loss is essentially zero, at the cost of no longer replaying the reporter's literal config.
2. **Capture evidence first**: upload the birdnet session log as an artifact for this lane and/or arm the faulthandler watchdog *below* the pytest timeout for this one test, so the next occurrence yields a stack instead of another silent `node down`.
3. **Leave it as a canary** and eat the occasional red macOS job.

Option 1 is the pragmatic fix; option 2 is worth it only if we want to know the true root cause before hiding it. Happy to take either in a PR.

*(Analysis assembled with [Claude Code](https://claude.com/claude-code); the run links, timings, and dump contents above are all from the referenced CI artifacts.)*

**Comment by Josef-Haupt** [comment]
## Reproduction attempt: negative result

The hang did not reproduce in **10 full CI re-runs plus 4 fresh-commit runs**, including runs deliberately loaded to exceed the conditions of the original failures. Details, so nobody repeats this work:

| phase | what ran | hangs |
| --- | --- | --- |
| 10 re-runs of `73bf241` | 1 instance/lane | 0 |
| fresh run 1 (`3ba5f8e`) | 5 instances/lane | 0 |
| fresh run 2 (`0e679a9`) | 5 instances/lane | 0 |
| fresh run 4 (`89ad7ca`) | 5 instances/lane, **18/18 jobs green** | 0 |

The diagnostic branch ([#69](https://github.com/birdnet-team/birdnet/pull/69)) parametrizes `test_issue_29` into five instances that xdist runs concurrently — each spawning the full default pipeline (`n_workers=None` → one TF worker per physical core, plus the perf tracker). In the final run every lane executed all five (`ubuntu-24.04/3.12` ran 15, across its three tox envs) and none wedged.

### Why several earlier attempts were worthless

`tox` runs `pytest -m 'not repro and load_model'` **first**. When that phase fails the env aborts and the general phase never runs — so `test_issue_29` executes zero times. That is exactly what happened on several lanes: transient download failures killed the `load_model` phase and silently deleted the coverage we were trying to sample. Verified on `ubuntu-24.04-arm/3.12` in fresh run 1: `test_issue_29` ran **0 times**.

Three separate infrastructure flakes were found and fixed while chasing this:

- `brew update` fetching every preinstalled tap (#67)
- `choco install libsndfile` (#70)
- the shared download helper making exactly one attempt (#71)

Fresh run 4 — the first run with all three in place — is also the first fully green run of this branch.

### What this means for the diagnosis

The trigger remains unidentified: both original failures were *first runs of fresh commits* in the evening (UTC), and nothing since has recreated the conditions. Runner-fleet contention at that time of day is plausible but not something we can steer.

**What is identified, and is not macOS-specific, is the amplifier.** The pipeline parent has no child-liveness monitoring on the happy path:

- `Consumer._run_main_loop` waits for one `None` sentinel per worker, with `cancel_event` as the only exit
- `ProcessManager.wait_until_all_finished()` blocks on finish signals with `timeout=None`
- `ProcessManager.join()` calls `p.join()` with no timeout

So if any spawned child dies or stalls during startup — OOM kill, jetsam, a crashed TF import, anything — nothing ever fires and the parent waits **forever**, silently. That matches the observed evidence exactly: no output whatsoever after the test's logstart line.

This is worth fixing on its own merits, independently of what triggers it in CI. For a user it is the difference between "birdnet raised an error explaining a worker died" and "birdnet froze and I killed it". A child that dies is currently indistinguishable from a child that is working.

### Suggested direction

Give the parent's waits a liveness check rather than an unconditional block: while waiting on the sentinels/finish signals, periodically verify every child process is still alive (`Process.is_alive()` / `exitcode`), and if one has died without signalling, set `cancel_event` and raise an error naming the process and its exit code. The `Consumer` loop already polls at 1 s (`self._queue.get(timeout=1.0)`), so the hook exists — it just does not check anything except `cancel_event` today.

Note that `test_cancel_teardown_v2_4.py` uses a 120 s deadline, so even a 50 s stall passes CI silently; a liveness check would surface these as errors instead of timeouts.

### Recommendation

Close the reproduction hunt, keep this issue open for the amplifier fix, and let [#69](https://github.com/birdnet-team/birdnet/pull/69) close unmerged (its forensics probe can be resurrected from the branch if the hang returns). If it does return, the probe now writes snapshots at 120/300/480 s — before pytest-timeout destroys the evidence — so a single recurrence would give us the stuck frames.

*(Investigation performed with [Claude Code](https://claude.com/claude-code). All run links and counts above are from the referenced CI artifacts and logs.)*

**Comment by Josef-Haupt** [comment]
Closing as not reproducible for now.

The hang did not recur in 10 CI re-runs plus 4 fresh-commit runs, including runs deliberately loaded with five concurrent instances of the test per lane — details in the comment above. Both original occurrences were first runs of fresh commits in the evening (UTC), and nothing since has recreated the conditions.

Several things found along the way have been fixed or are tracked separately, so if this recurs the ground is better prepared than it was:

- #72 — a pipeline child that dies is now reported instead of hanging the run
- #75 / #78 — a run producing no predictions no longer hangs the progress dispatcher
- #73 — a worker killed mid-batch deadlocks the survivors on POSIX (open, documented)
- #77 — the teardown drain can block on a truncated frame (open)

The branch `diag/issue29-hang-forensics` still holds a probe that snapshots child process state, memory and py-spy stacks at 120/300/480 s — i.e. before pytest-timeout kills the worker and destroys the evidence. One recurrence with that branch active would give the stuck frames.

Reopen if it comes back.
